File: src/move.ts
import { Board, Scores, Side } from "./interfaces";
export const getOtherSide = (side: Side) => (side === "top" ? "bot" : "top");

const hasMoves = (board: Board, side: Side) =>
  board[side].some(hole => hole > 0);

const newBoard: Board = {
  currentTurn: "top",
  top: [7, 7, 7, 7, 7, 7, 7],
  bot: [7, 7, 7, 7, 7, 7, 7],
  scores: {
    top: 0,
    bot: 0,
  },
  gameOver: false,
  turnsPlayed: 0,
};

export const newGame = (board: Board) => (board.gameOver ? newBoard : board);

const getWinner = ({ top, bot }: Scores) =>
  top > bot ? "top" : bot > top ? "bot" : "draw";

export class BoardMove {
  board: Board;
  side: Side;

  constructor(board: Board, side: Side) {
    this.board = board;
    this.side = side;
  }

  makeMove(holeIdx: number) {
    const outOfRange = holeIdx < 0 || holeIdx > 6;
    const invalidTurn = this.side !== this.board.currentTurn;

    let shells = this.board[this.side][holeIdx];
    if (!shells || outOfRange || invalidTurn || this.board.gameOver)
      return this.board;

    this.board[this.side][holeIdx] = 0;
    let startingIdx = holeIdx + 1;
    let currentSide = this.side;
    let extraMove = false;
    while (shells > 0) {
      shells = this.distributeStones(startingIdx, shells, currentSide);
      if (shells && currentSide === this.side) {
        ++this.board.scores[this.side];
        --shells;
        if (shells === 0) extraMove = true;
      }
      currentSide = getOtherSide(currentSide);
      startingIdx = 0;
    }

    return this.evaluateAfterMove(extraMove);
  }

  distributeStones(startingIdx: number, shells: number, currentSide: Side) {
    const otherSide = getOtherSide(this.side);
    let shellCount = shells;
    for (let i = startingIdx; shellCount && i < 7; ++i) {
      // last shell mechanics
      const lastShell = shellCount === 1;
      const nextPitEmpty = !this.board[currentSide][i];
      const otherPitEmpty = !this.board[otherSide][6 - i];
      const canCapture =
        nextPitEmpty && !otherPitEmpty && currentSide === this.side;

      if (lastShell && !nextPitEmpty) {
        shellCount = ++this.board[currentSide][i];
        this.board[currentSide][i] = 0;
      } else if (lastShell && canCapture) {
        this.board.scores[this.side] += ++this.board[otherSide][6 - i];
        this.board[otherSide][6 - i] = 0;
        --shellCount;
      } else {
        --shellCount;
        ++this.board[currentSide][i];
      }
    }

    return shellCount;
  }

  evaluateAfterMove(extraMove: boolean) {
    const otherSide = getOtherSide(this.side);
    const playerHasMoves = hasMoves(this.board, this.side);
    const enemyHasMoves = hasMoves(this.board, otherSide);

    const moveAgain = (extraMove && playerHasMoves) || !enemyHasMoves;
    this.board.currentTurn = moveAgain ? this.side : otherSide;

    if (!moveAgain) ++this.board.turnsPlayed;

    if (!(playerHasMoves || enemyHasMoves)) {
      this.board.gameOver = true;
      this.board.currentTurn = getWinner(this.board.scores);
      ++this.board.turnsPlayed;
    }

    return this.board;
  }
}

// AI stuff

const getRandomIdx = (arr: number[]) => Math.floor(Math.random() * arr.length);

const randomMove = (moves: number[]) => moves[getRandomIdx(moves)];

function getPossibleMoves(row: number[]) {
  const moves: number[] = [];
  const wor = row.reverse();
  wor.forEach((num, i) => {
    if (num) moves.push(i);
  });
  return moves;
}

function chooseMove({ board, side }: BoardMove) {
  const thisRow = board[side];
  /*
  TO-DO: try to enforce capturing first before trying to make a random move
  const otherSide = getOtherSide(side);
  const otherRow = board[otherSide];
  */

  const possibleMoves = getPossibleMoves(thisRow);

  // guarantees an extra turn
  for (let idx of possibleMoves) if (thisRow[idx] + idx === 7) return idx;

  // last resort
  return randomMove(possibleMoves);
}

export function makeAIMove(board: Board) {
  if (board.gameOver) return board;
  // currentSide will never be draw
  const currentSide = <Side>board.currentTurn;
  const boardMove = new BoardMove(board, currentSide);

  let isTurn = boardMove.board.currentTurn === currentSide;
  while (isTurn && !boardMove.board.gameOver) {
    const holeIdx = chooseMove(boardMove);
    boardMove.makeMove(holeIdx);
    isTurn = boardMove.board.currentTurn === currentSide;
  }

  return boardMove.board;
}

File: cmd/sungka/move.go
package main

import (
	"math/rand"
)

func getOtherSide(s Side) Side {
	if s == Top {
		return Bot
	} else {
		return Top
	}
}

func hasMoves(b Board, side Side) bool {
	if side == Top {
		for _, hole := range b.top {
			if hole > 0 {
				return true
			}
		}
	} else if side == Bot {
		for _, hole := range b.bot {
			if hole > 0 {
				return true
			}
		}

	}

	return false
}

func newBoard() Board {
	return Board{
		currentTurn: Top,
		top:         []int{7, 7, 7, 7, 7, 7, 7},
		bot:         []int{7, 7, 7, 7, 7, 7, 7},
		scores: Scores{
			top: 0,
			bot: 0,
		},
		gameOver:    false,
		turnsPlayed: 0,
	}
}

func newGame(board Board) Board {
	if board.gameOver {
		return newBoard()
	} else {
		return board
	}
}

func getWinner(s Scores) Side {
	if s.top > s.bot {
		return Top
	} else if s.bot > s.top {
		return Bot
	} else {
		return Draw
	}
}

type BoardMove struct {
	board Board
	side  Side
}

func newBoardMove(b Board, s Side) BoardMove {
	return BoardMove{b, s}
}

func (b *BoardMove) makeMove(holeIdx int) Board {
	outOfRange := holeIdx < 0 || holeIdx > 6
	invalidTurn := b.side != b.board.currentTurn
	shells := (*b.getSidePits(b.side))[holeIdx]
	if shells == 0 || outOfRange || invalidTurn || b.board.gameOver {
		return b.board
	}

	(*b.getSidePits(b.side))[holeIdx] = 0
	startingIdx := holeIdx + 1
	currentSide := b.side
	extraMove := false

	for shells > 0 {
		shells = b.distributeStones(startingIdx, shells, currentSide)
		if shells > 0 && currentSide == b.side {
			if b.side == Top {
				b.board.scores.top++
			} else if b.side == Bot {
				b.board.scores.bot++
			}
			shells--
			extraMove = shells == 0
		}
		currentSide = getOtherSide(currentSide)
		startingIdx = 0
	}
	return b.evaluateAfterMove(extraMove)
}

func getSidePits(b *Board, s Side) *[]int {
	if s == Top {
		return &b.top
	} else if s == Bot {
		return &b.bot
	}
	return nil
}

func (b *BoardMove) getSidePits(s Side) *[]int {
	return getSidePits(&b.board, s)
}

func getSideScore(b *Board, s Side) *int {
	if s == Top {
		return &b.scores.top
	} else if s == Bot {
		return &b.scores.bot
	}
	return nil
}

func (b *BoardMove) getSideScore(s Side) *int {
	return getSideScore(&b.board, s)
}

func (b *BoardMove) distributeStones(startingIdx, shells int, currentSide Side) int {
	otherSide := getOtherSide(b.side)
	shellCount := shells
	for i := startingIdx; shellCount > 0 && i < 7; i++ {
		// last shell mechanics
		lastShell := shellCount == 1
		nextPitEmpty := (*b.getSidePits(currentSide))[i] == 0
		otherPitEmpty := (*b.getSidePits(otherSide))[6-i] == 0
		canCapture := nextPitEmpty && !otherPitEmpty && currentSide == b.side

		if lastShell && !nextPitEmpty {
			(*b.getSidePits(currentSide))[i]++
			shellCount = (*b.getSidePits(currentSide))[i]
			(*b.getSidePits(currentSide))[i] = 0
		} else if lastShell && canCapture {
			(*b.getSidePits(otherSide))[6-i]++
			*b.getSideScore(b.side) += (*b.getSidePits(otherSide))[6-i]
			shellCount--
		} else {
			shellCount--
			(*b.getSidePits(currentSide))[i]++
		}
	}
	return shellCount
}

func (b *BoardMove) evaluateAfterMove(extraMove bool) Board {
	otherSide := getOtherSide(b.side)
	playerHasMoves := hasMoves(b.board, b.side)
	enemyHasMoves := hasMoves(b.board, otherSide)
	moveAgain := (extraMove && playerHasMoves) || !enemyHasMoves
	if moveAgain {
		b.board.currentTurn = b.side
	} else {
		b.board.currentTurn = otherSide
	}

	if !moveAgain {
		b.board.turnsPlayed++
	}

	if !(playerHasMoves || enemyHasMoves) {
		b.board.gameOver = true
		b.board.currentTurn = getWinner(b.board.scores)
		b.board.turnsPlayed++
	}

	return b.board
}

// AI Stuff
func randomMove(moves []int) int {
	return moves[rand.Intn(len(moves))]
}

func getPossibleMoves(row []int) []int {
	moves := []int{}
	for i := len(row) - 1; i >= 0; i-- {
		if row[i] > 0 {
			moves = append(moves, i)
		}
	}
	return moves
}

func chooseMove(b BoardMove) int {
	thisRow := *b.getSidePits(b.side)
	/*
		TO-DO: try to enforce capturing first before trying to make a random move
		otherSide := getOtherSide(b.side)
		otherRow := *b.getSidePits(otherSide)
	*/
	possibleMoves := getPossibleMoves(thisRow)
	for _, i := range possibleMoves {
		if thisRow[i]+i == 7 {
			return i
		}
	}
	return randomMove(possibleMoves)
}

func makeAIMove(board Board) Board {
	if board.gameOver {
		return board
	}
	currentSide := board.currentTurn
	b := newBoardMove(board, currentSide)
	isTurn := b.board.currentTurn == currentSide
	for isTurn && !b.board.gameOver {
		holeIdx := chooseMove(b)
		b.makeMove(holeIdx)
		isTurn = b.board.currentTurn == currentSide
	}

	return b.board
}
